common/web/keyboard-processor/src/keyboards/activeLayout.ts

```diff
@@ -428,16 +438,17 @@ namespace com.keyman.keyboards {
           if(key == null) {
             keys.length = keys.length-1;
           } else {
-            var kw, kp;
-            kw = (typeof key['width'] == 'string' && key['width'] != '') ? parseInt(key['width'],10) : ActiveKey.DEFAULT_KEY_WIDTH;
-            if(isNaN(kw) || kw == 0) kw = ActiveKey.DEFAULT_KEY_WIDTH;
-            key['width'] = kw.toString();
-            kp = (typeof key['pad'] == 'string' && key['pad'] != '') ? parseInt(key['pad'],10) : ActiveKey.DEFAULT_PAD;
-            if(isNaN(kp) || kp == 0) kp = ActiveKey.DEFAULT_PAD;  // KMEW-119
-            key['pad'] = kp.toString();
-            width += kw + kp;
-            //if(typeof key['width'] == 'string' && key['width'] != '') width += parseInt(key['width'],10); else width += DEFAULT_KEY_WIDTH;
-            //if(typeof key['pad'] == 'string' && key['pad'] != '') width += parseInt(key['pad'],10); else width += 5;
+            // Note:  the key has not yet been 'polyfilled'; these _may_ be strings.
+            // We coerce their values to numbers permanently here.
+            if(typeof key.width == 'string') {
+              key.width = parseInt(key.width, 10); // '' => NaN.
+            }
+            width += key.width = (isNaN(key.width as number) ? 0 : key.width) || ActiveKey.DEFAULT_KEY_WIDTH;
```

Comment:
Is this the first place where key.width is tested? My thinking is that after this, `key.width` should be a guaranteed good value, which simplifies all future references (e.g. lines 356-374 which otherwise will flub on bad data). Also, I think `''` should be equivalent to default width, rather than `0`.

```suggestion
            if(typeof key.width == 'string') {
              key.width = parseInt(key.width, 10); // '' => NaN.
            }
            if(typeof(key.width) != 'number' || isNaN(key.width)) {
              key.width = ActiveKey.DEFAULT_KEY_WIDTH;
            }
            width += key.width;
```

There might be better ways to solve this, so I am happy if you want to do it another way.

Same should be done to `key.pad`.

I'd also be quite open to the idea of a more direct refactor, moving all cleanup out of this function and into a load phase function, which would make it easier to review and unit test.

Reply:
> Also, I think '' should be equivalent to default width, rather than 0.

Well, fortunately, the version currently in this PR does exactly this.

Line 444:  `''` is "coerced" to NaN.

Line 446:
1.  `isNaN` is used to replace `NaN` with `0`.
2. `0 || DEFAULT_KEY_WIDTH` results in the latter value.

Note that the original code being replaced always replaced a value of `0` with `DEFAULT_KEY_WIDTH` as well... something that is lacking from your suggestion.

Best-match original line, since GitHub isn't showing it in this comment's view:

```typescript
if(isNaN(kw) || kw == 0) kw = ActiveKey.DEFAULT_KEY_WIDTH;
```

(Where `kw` is short for `keyWidth`)

Reply:
> Is this the first place where key.width is tested? My thinking is that after this, `key.width` should be a guaranteed good value, which simplifies all future references (e.g. lines 356-374 which otherwise will flub on bad data).

This is exactly the solution I've landed on and have already used to do said simplification.